README.md
# Syslog File Converter (syslog_fc)

[![Build Status](https://img.shields.io/travis/namedun/syslog_fc/master.svg)](https://travis-ci.org/namedun/syslog_fc) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/f6090a0a38164300a7f95b709b704bd4)](https://www.codacy.com/app/namedun/syslog_fc?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=namedun/syslog_fc&amp;utm_campaign=Badge_Grade)

This utility is designed to convert syslog files to various other text formats, such as CSV, HTML or JSON (full list of the supported output formats are listed below in the "[Supported Output Formats](#supported-output-formats)" section).

## Build and Install

Run CMake (notice the dot at the end):

```shell
$ cmake .
```

Run Make:

```shell
$ make
```

This will produce binary `syslogfc` in the current directory. To install this binary to `/usr/bin` run command as `root` user:

```shell
# make install
```

## Usage

Usage syntax:
```shell
syslogfc [options] <input-file>
```

*   `[options]` is a one or more additional optional options that are described in the "[Options](#options)" section.
*   `<input-file>` is the path to the syslog file to be converted. Input syslog file path should not be specified if selected the input from standard input (stdin) using additional option `--stdin` (see the "[Options](#options)" section).

### Options

#### `-h`, `--help`

Show help and usage text.

#### `-s`, `--stdin`

Use standard input (stdin) for input data. When this option is specified, no `<input-file>` is required.

#### `-f <format>`, `--format=<format>`

Output format selection. Available output formats listed int the "[Supported Output Formats](#supported-output-formats)" section.

Default: `plain`.

#### `-e <spec>`, `--entry-spec=<spec>`

Syslog entry fields specification. This specification determines the order and the presence of certain message fields in the system log files.

Available syslog entry specificators are listed below:

| Specificator | Description                                                                                                                                                                                                                              |
| ------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `%I`         | Numeric identifier                                                                                                                                                                                                                       |
| `%T`         | Timestamp. For parsing this field used timestamp parsing format specification, which can be specified by option `--ts-parse-spec`.                                                                                                       |
| `%K`         | Kernel time.                                                                                                                                                                                                                             |
| `%H`         | Hostname.                                                                                                                                                                                                                                |
| `%F`         | Facility. Valid facilities are `auth`, `authpriv`, `cron`, `daemon`, `ftp`, `kern`, `lpr`, `mail`, `mark`, `news`, `security`, `syslog`, `user`, `uucp`, `local0`, `local1`, `local2`, `local3`, `local4`, `local5`, `local6`, `local7`. |
| `%P`         | Priority. Valid priorities are `alert`, `crit`, `debug`, `emerg`, `err`, `error`, `info`, `none`, `notice`, `panic`, `warn`, `warning`.                                                                                                  |
| `%G`         | Tag.                                                                                                                                                                                                                                     |
| `%M`         | Message.                                                                                                                                                                                                                                 |

One or more special modifiers can be specified between the `%` character and the specifier character, for example `%_@P`. The modifiers allows you to control the process of parsing or displaying (output) the values of the corresponding fields. For example, with the modifiers you can prevent the trimming spaces for a specific field, or disable value validation for it.

Available modifiers:

| Modifier | Description                                                                         |
| -------- | ----------------------------------------------------------------------------------- |
| `_`      | Do not trim spaces at the beginning. No effect for timestamp and identifier fields. |
| `@`      | Do not validate parsed value.                                                       |
| `!`      | Do not use this field in the output.                                                |

Default entry fields specification is `%T %F.%P %G: %_M`, which corresponds to the syslog file entries in the form:
```plain
<Timestamp> <Facility>.<Priority> <Tag>: <Message>
```

For example:
```plain
Mon Jun 24 18:12:50 2019 kern.info kernel: br-lan: port 1(sw1p1) entered blocking state
└┬─────────────────────┘ └┬─┘ └┬─┘ └┬───┘  └┬─────────────────────────────────────────┘
 │                        │    │    │       │
 │                        │    │    │       └──> Message
 │                        │    │    └──────────> Tag
 │                        │    └───────────────> Priority
 │                        └────────────────────> Facility
 └─────────────────────────────────────────────> Timestamp
```

#### `-p <spec>`, `--ts-parse-spec=<spec>`

Timestamp parsing format specification.

The format specification `<spec>` is a string and may contain special character sequences called conversion specifications, each of which is introduced by a `%` character and terminated by some other character known as a conversion specifier character. All other character sequences are ordinary character sequences.

All available conversion specifier characters are listed at the `strptime()` function manual page (<https://linux.die.net/man/3/strptime>).

Note, that if the parameter value contains spaces, it must be enclosed in quotes. For example:
```shell
$ syslogfc --ts-parse-spec="%b %d %H:%M:%S" /path/to/syslog.log
```

Default: `%a %b %d %H:%M:%S %Y`.

#### `-o <spec>`, `--ts-output-spec=<spec>`

Timestamp output format specification.

The format specification `<spec>` is a string and may contain special character sequences called conversion specifications, each of which is introduced by a `%` character and terminated by some other character known as a conversion specifier character. All other character sequences are ordinary character sequences.

All available conversion specifier characters are listed at the `strftime()` function manual page (<https://linux.die.net/man/3/strftime>).

Note, that if the parameter value contains spaces, it must be enclosed in quotes. For example:
```shell
$ syslogfc --ts-output-spec="%d.%m.%Y %H:%M:%S" /path/to/syslog.log
```

Default format for output is not specified. If the output format is not specified or specified as empty string, then the time will be output as UNIX timestamp.

#### `-d <delimieter>`, `--csv-delimeter=<delimeter>`

Delimeter for `csv` (CSV) output format.

Default: `,` (comma).

#### `-x <prefix>`, `--html-class-prefix=<prefix>`

Prefix for HTML classes for `html` output format.

Default: `syslog-`.

#### `-c <on|off>`, `--html-cell-classes=<on|off>`

Output HTML classes for each table cell for `html` output format.

Default: `off`.

## Supported Output Formats

| Format     | Description                            |
| ---------- | -------------------------------------- |
| `asciidoc` | AsciiDoc                               |
| `csv`      | CSV (Comma-Separated Values)           |
| `html`     | HTML (HyperText Markup Language) table |
| `json`     | JSON (JavaScript Object Notation)      |
| `md`       | Markdown table                         |
| `plain`    | Plain text format (for testing)        |

## Examples

### JSON

Input syslog file `example1.log`:

```plain
Mon Jun 24 18:12:50 2019 kern.info kernel: br-lan: port 1(sw1p1) entered blocking state
Mon Jun 24 18:12:50 2019 daemon.info mstpd[1882]: set_if_up: Port "sw1p1" : up
Mon Jun 24 18:12:50 2019 daemon.notice netifd: bridge "br-lan" link is up
Mon Jun 24 18:13:07 2019 daemon.err modprobe: xt_connmark is already loaded
```

Convert to JSON:

```shell
$ syslogfc --format=json example1.log
[{"timestamp":"1561389170","facility":"kern","priority":"info","tag":"kernel","message":"br-lan: port 1(sw1p1) entered blocking state"},{"timestamp":"1561389170","facility":"daemon","priority":"info","tag":"mstpd[1882]","message":"set_if_up: Port \"sw1p1\" : up"},{"timestamp":"1561389170","facility":"daemon","priority":"notice","tag":"netifd","message":"bridge \"br-lan\" link is up"},{"timestamp":"1561389187","facility":"daemon","priority":"err","tag":"modprobe","message":"xt_connmark is already loaded"}]
```

If you need formatted JSON output, you can use any external JSON formatting tool. For example:

```shell
$ syslogfc --format=json example1.log | python3 -m json.tool
[
    {
        "timestamp": "1561389170",
        "facility": "kern",
        "priority": "info",
        "tag": "kernel",
        "message": "br-lan: port 1(sw1p1) entered blocking state"
    },
    {
        "timestamp": "1561389170",
        "facility": "daemon",
        "priority": "info",
        "tag": "mstpd[1882]",
        "message": "set_if_up: Port \"sw1p1\" : up"
    },
    {
        "timestamp": "1561389170",
        "facility": "daemon",
        "priority": "notice",
        "tag": "netifd",
        "message": "bridge \"br-lan\" link is up"
    },
    {
        "timestamp": "1561389187",
        "facility": "daemon",
        "priority": "err",
        "tag": "modprobe",
        "message": "xt_connmark is already loaded"
    }
]
```

### CSV

Input syslog file `example2.log`:

```plain
24.06.2019 18:12:50 kern.info kernel: br-lan: port 1(sw1p1) entered blocking state
24.06.2019 18:12:50 daemon.info mstpd[1882]: set_if_up: Port "sw1p1" : up
24.06.2019 18:12:50 daemon.notice netifd: bridge "br-lan" link is up
24.06.2019 18:13:07 daemon.err modprobe: xt_connmark is already loaded
```

Note that the timestamp format is different from the previous example.

If we try to convert the file using the default timestamp parsing format, we get parsing errors:
```shell
$ syslogfc --format=csv example2.log
1: Failed to parse timestamp (-84)
2: Failed to parse timestamp (-84)
3: Failed to parse timestamp (-84)
4: Failed to parse timestamp (-84)
```

For proper conversion, you must specify the correct format:
```shell
$ syslogfc -p "%d.%m.%Y %H:%M:%S" --format=csv example2.log
"1561389170","kern","info","kernel","br-lan: port 1(sw1p1) entered blocking state"
"1561389170","daemon","info","mstpd[1882]","set_if_up: Port ""sw1p1"" : up"
"1561389170","daemon","notice","netifd","bridge ""br-lan"" link is up"
"1561389187","daemon","err","modprobe","xt_connmark is already loaded"
```

For example, let's also change the output time format:
```shell
$ syslogfc -p "%d.%m.%Y %H:%M:%S" -o "%A %d %B %Y" --format=csv example2.log
"Monday 24 June 2019","kern","info","kernel","br-lan: port 1(sw1p1) entered blocking state"
"Monday 24 June 2019","daemon","info","mstpd[1882]","set_if_up: Port ""sw1p1"" : up"
"Monday 24 June 2019","daemon","notice","netifd","bridge ""br-lan"" link is up"
"Monday 24 June 2019","daemon","err","modprobe","xt_connmark is already loaded"
```

## Changelog

See [CHANGELOG.md](CHANGELOG.md).

## License

This work is free. You can redistribute it and/or modify it under the
terms of the Do What The Fuck You Want To Public License, Version 2,
as published by Sam Hocevar. See <http://www.wtfpl.net/> for more details.

## Authors

*   Anton Kikin <mailto:a.kikin@tano-systems.com>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/0.9.0/).

All dates in this document are in `DD.MM.YYYY` format.

## [Unreleased][]

Initial release

[unreleased]: https://github.com/namedun/syslog_fc/tree/master


src/main.c
/*
 * Syslog File Converter
 * Copyright © 2019 Anton Kikin <a.kikin@tano-systems.com>
 *
 * This tool can convert syslog files into different text
 * formats, such as CSV, JSON or HTML.
 *
 * Default syslog file format:
 *   <timestamp> <facility>.<priority> <tag>: <message>
 *
 * Default syslog file format can be changed by option --entry-spec.
 *
 * Default <timestamp> format is "%a %b %d %H:%M:%S %Y" (strptime format).
 * Custom <timestamp> format can be specified by option --ts-parse-spec (-p).
 *
 * For more details see README.md.
 *
 * This work is free. You can redistribute it and/or modify it under the
 * terms of the Do What The Fuck You Want To Public License, Version 2,
 * as published by Sam Hocevar. See the COPYING file for more details.
 */

/**
 * @file
 * @brief Main source file
 *
 * @author Anton Kikin <a.kikin@tano-systems.com>
 */

#include <getopt.h>

#include <syslog_fc.h>

#include <fmt_plain.h>
#include <fmt_json.h>
#include <fmt_csv.h>
#include <fmt_md.h>
#include <fmt_html.h>
#include <fmt_asciidoc.h>

/**
 * @brief Output formats
 */
static const output_fmt_t *fmt_avail[] =
{
	&fmt_plain,
	&fmt_md,
	&fmt_csv,
	&fmt_json,
	&fmt_html,
	&fmt_asciidoc,
	NULL
};

/**
 * @brief Default configuration structure
 */
static const config_t default_config =
{
	.is_stdin          =  0,
	.output_fmt        = &fmt_plain,
	.entry_spec        = "%T %F.%P %G: %_M",
	.ts_parse_spec     = "%a %b %d %H:%M:%S %Y", /* Mon Jun 24 18:12:50 2019 */
	.ts_output_spec    = "",                     /* UNIX timestamp */
	.input_filename    =  NULL,
	.csv_delimeter     = ",",
	.html_class_prefix = "syslog-",
	.html_cell_classes =  0,
};

/**
 * @brief Global configuration structure
 */
config_t config = { 0 };

/**
 * @brief Short command line options list
 */
static const char *opts_str = "hf:e:sp:o:d:x:c:";

/**
 * @brief Long command line options list
 */
static const struct option opts[] =
{
	{ .name = "help",              .val = 'h' },
	{ .name = "format",            .val = 'f', .has_arg = 1 },
	{ .name = "stdin",             .val = 's' },
	{ .name = "entry-spec",        .val = 'e', .has_arg = 1 },
	{ .name = "ts-parse-spec",     .val = 'p', .has_arg = 1 },
	{ .name = "ts-output-spec",    .val = 'o', .has_arg = 1 },
	{ .name = "csv-delimeter",     .val = 'd', .has_arg = 1 },
	{ .name = "html-class-prefix", .val = 'x', .has_arg = 1 },
	{ .name = "html-cell-classes", .val = 'c', .has_arg = 1 },
	{ 0 }
};

/**
 * Display program usage help
 */
static void display_usage(void)
{
	int i;

	fprintf(stdout,
		"\n"
		"Syslog File Converter version " SYSLOG_FC_VERSION "\n"
		"Copyright (c) 2019 Anton Kikin <a.kikin@tano-systems.com>\n"
		"\n"
		"Usage: syslog_fc [options] <input-file>\n"
		"\n"
		"Options:\n"
		"  -h, --help\n"
		"        Show this help text.\n"
		"\n"
		"  -s, --stdin\n"
		"        Read data from stdin instead of file.\n"
		"\n"
		"  -f, --format <format>\n"
		"        Select output format.\n"
		"\n"
		"        Available output formats:\n"
	);

	/* Display available formats */
	for (i = 0; fmt_avail[i]; i++)
	{
		fprintf(stdout, "%12s%-8s - %s\n",
			"", /* left indentation */
			fmt_avail[i]->name,
			fmt_avail[i]->description
		);
	}

	fprintf(stdout,
		"\n"
		"        Default: \"%s\"\n"
		"\n"
		"  -e, --entry-spec <spec>\n"
		"        Syslog entry fields specification.\n"
		"        See README.md for details.\n"
		"\n"
		"        Allowed format specificators:\n"
		"            %%T - Timestamp\n"
		"            %%K - Kernel time\n"
		"            %%H - Hostname\n"
		"            %%F - Facility\n"
		"            %%P - Priority\n"
		"            %%G - Tag\n"
		"            %%M - Message\n"
		"\n"
		"        Default: \"%s\"\n"
		"\n"
		"  -p, --ts-parse-spec <format>\n"
		"        Timestamp format specification for parsing.\n"
		"        See 'man strptime' for available specificators description.\n"
		"\n"
		"        Default: \"%s\"\n"
		"\n"
		"  -o, --ts-output-spec <format>\n"
		"        Timestamp format specification for output.\n"
		"        Keep empty for output time as UNIX timestamp.\n"
		"        See 'man strftime' for available specificators description.\n"
		"\n"
		"        Default: \"%s\"\n"
		"\n"
		"  -d, --csv-delimeter <delimeter>\n"
		"        Specifiy delimeter for CSV output format.\n"
		"\n"
		"        Default: \"%s\"\n"
		"\n"
		"  -x, --html-class-prefix <prefix>\n"
		"        Prefix for HTML classes.\n"
		"\n"
		"        Default: \"%s\"\n"
		"\n"
		"  -c, --html-cell-classes <on|off>\n"
		"        Add HTML classes for each table cell.\n"
		"\n"
		"        Default: \"%s\"\n"
		"\n",
		default_config.output_fmt->name,
		default_config.entry_spec,
		default_config.ts_parse_spec,
		default_config.ts_output_spec ? default_config.ts_output_spec : "",
		default_config.csv_delimeter,
		default_config.html_class_prefix,
		default_config.html_cell_classes ? "on" : "off"
	);
}

/**
 * Parse command line arguments into @ref config global structure
 *
 * @param[in] argc  Number of arguments
 * @param[in] argv  Array of the pointers to the arguments
 *
 * @return 0 on success
 * @return <0 on error
 */
static int cli_args(int argc, char *argv[])
{
	int opt;

	while((opt = getopt_long(argc, argv, opts_str, opts, NULL)) != EOF)
	{
		switch(opt)
		{
			case '?':
			{
				/* Invalid option */
				return -EINVAL;
			}

			case 'h': /* --help */
			{
				display_usage();
				exit(0);
			}

			case 's': /* --stdin */
			{
				config.is_stdin = 1;
				break;
			}

			case 'f': /* --format */
			{
				int i;
				const output_fmt_t *new_output_fmt = NULL;

				for (i = 0; fmt_avail[i]; i++)
				{
					if (!strcmp(optarg, fmt_avail[i]->name))
					{
						new_output_fmt = fmt_avail[i];
						break;
					}
				}

				if (!new_output_fmt)
				{
					fprintf(stderr, "%s: invalid format '%s'\n",
						argv[0], optarg);

					return -EINVAL;
				}

				config.output_fmt = new_output_fmt;
				break;
			}

			case 'e': /* --entry-spec */
			{
				config.entry_spec = optarg;
				break;
			}

			case 'd': /* --csv-delimeter */
			{
				config.csv_delimeter = optarg;
				break;
			}

			case 'p': /* --ts-parse-spec */
			{
				config.ts_parse_spec = optarg;
				break;
			}

			case 'o': /* --ts-output-spec */
			{
				config.ts_output_spec = optarg;
				break;
			}

			case 'x': /* --html-class-prefix */
			{
				config.html_class_prefix = optarg;
				break;
			}

			case 'c': /* --html-cell-classes */
			{
				if ((strcmp(optarg, "on") == 0) ||
				    (strcmp(optarg, "true") == 0) ||
				    (strcmp(optarg, "1") == 0))
					config.html_cell_classes = 1;
				else
					config.html_cell_classes = 0;

				break;
			}

			default:
				break;
		}
	}

	if ((argc == optind) && !config.is_stdin)
	{
		fprintf(stderr, "%s: input file is not specified\n", argv[0]);
		return -EINVAL;
	}
	else
	{
		if (!config.is_stdin)
		{
			config.input_filename = argv[optind];
		}
		else if (argc > optind)
		{
			fprintf(stderr,
				"%s: can't specify both stdin and input file\n",
				argv[0]
			);

			return -EINVAL;
		}
	}

	return 0;
}

/**
 * Convert syslog file into other text format
 *
 * @param[in] input Pointer to the input syslog file structure
 *
 * @return 0 on success
 * @return <0 on error
 */
static int convert_syslog(FILE *input)
{
	int ret = 0;
	syslog_entry_t entry;

	char *buffer;
	size_t buffer_size = SYSLOG_BUFFER_SIZE;

	unsigned int parsed_n = 0;
	unsigned int line_n = 0;

	buffer = malloc(buffer_size);
	if (!buffer)
	{
		fprintf(stderr,
			"Failed to allocate memory for line buffer\n");

		return -ENOMEM;
	}

	ret = syslog_entry_init(&entry, config.entry_spec);
	if (ret)
	{
		fprintf(stderr,
			"Syslog entry initialization failed (%d)\n", ret);

		free(buffer);
		return ret;
	}

	if (config.output_fmt->fn_output_start)
		config.output_fmt->fn_output_start(&entry);

	while (1)
	{
		int status;
		size_t buffer_offset = 0;
		size_t line_len = 0;

		line_n++;

		while (1)
		{
			char *new_buffer;
			char *line;

			line = fgets(buffer + buffer_offset,
				buffer_size - buffer_offset, input);

			if (!line)
				break;

			line_len += strnlen(line, buffer_size);

			if ((buffer[line_len - 1] == '\r') ||
			    (buffer[line_len - 1] == '\n'))
				break;

			/* Increase line buffer size */
			buffer_size += SYSLOG_BUFFER_SIZE;
			if (buffer_size > SYSLOG_MAX_BUFFER_SIZE)
			{
				fprintf(stderr,
					"line %u: Line buffer size limit (%zu) reached\n",
					line_n, (size_t)SYSLOG_MAX_BUFFER_SIZE);

				free(buffer);
				return -EINVAL;
			}

			new_buffer = realloc(buffer, buffer_size);
			if (!new_buffer)
			{
				fprintf(stderr,
					"line %u: Failed to reallocate memory for line buffer "
					"(%zu -> %zu)\n",
					line_n,
					buffer_size - SYSLOG_BUFFER_SIZE,
					buffer_size);

				free(buffer);
				return -ENOMEM;
			}

			buffer = new_buffer;
			buffer_offset = line_len;
		}

		if (!line_len) /* EOF */
			break;

		status = syslog_entry_parse(
			&entry, line_n, buffer);

		if (!status)
		{
			entry.num = ++parsed_n;

			if (config.output_fmt->fn_output_entry)
				config.output_fmt->fn_output_entry(&entry);
		}
	}

	if (config.output_fmt->fn_output_end)
		config.output_fmt->fn_output_end(&entry);

	syslog_entry_destroy(&entry);

	free(buffer);
	return ret;
}

/**
 * Program start point
 *
 * @param[in] argc  Number of arguments
 * @param[in] argv  Array of the pointers to the arguments
 *
 * @return 0 on success
 * @return <0 on error
 */
int main(int argc, char *argv[])
{
	int ret = 0;
	FILE *input;

	memcpy(&config, &default_config, sizeof(config));

	if (cli_args(argc, argv))
	{
		display_usage();
		return -EINVAL;
	}

	if (config.is_stdin)
		input = stdin;
	else
	{
		input = fopen(config.input_filename, "rb");
		if (!input)
		{
			fprintf(stderr, "%s: could not open file '%s'\n",
				argv[0], config.input_filename);

			return -ENODEV;
		}
	}

	ret = convert_syslog(input);

	if (!config.is_stdin)
		fclose(input);

	return ret;
}
